pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go

```diff
@@ -1841,242 +1842,312 @@ func getNumberOfPodInterfaces(spec *v1.VirtualMachineInstanceSpec) int {
 	return nPodInterfaces
 }
 
-func validateDisks(field *k8sfield.Path, disks []v1.Disk) []metav1.StatusCause {
+func validateDiskName(field *k8sfield.Path, idx int, disks []v1.Disk) []metav1.StatusCause {
 	var causes []metav1.StatusCause
-	nameMap := make(map[string]int)
-
-	for idx, disk := range disks {
-		// verify name is unique
-		otherIdx, ok := nameMap[disk.Name]
-		if !ok {
-			nameMap[disk.Name] = idx
-		} else {
+	for otherIdx, disk := range disks {
+		if otherIdx < idx && disk.Name == disks[idx].Name {
 			causes = append(causes, metav1.StatusCause{
 				Type:    metav1.CauseTypeFieldValueInvalid,
 				Message: fmt.Sprintf("%s and %s must not have the same Name.", field.Index(idx).String(), field.Index(otherIdx).String()),
 				Field:   field.Index(idx).Child("name").String(),
 			})
 		}
+	}
+	return causes
+}
 
-		// Verify only a single device type is set.
-		deviceTargetSetCount := 0
-		var diskType string
-		var bus v1.DiskBus
-		if disk.Disk != nil {
-			deviceTargetSetCount++
-			diskType = "disk"
-			bus = disk.Disk.Bus
-		}
-		if disk.LUN != nil {
-			deviceTargetSetCount++
-			diskType = "lun"
-			bus = disk.LUN.Bus
-		}
-		if disk.CDRom != nil {
+func validateDeviceTarget(field *k8sfield.Path, idx int, disk v1.Disk) []metav1.StatusCause {
+	var causes []metav1.StatusCause
+	deviceTargetSetCount := 0
+	for _, target := range []interface{}{disk.Disk, disk.LUN, disk.CDRom} {
+		if !reflect.ValueOf(target).IsNil() {
```

Comment:
Can it be done without reflection, just by checking if the interface is nil?

Reply:
I had initially tried it the following way
```
if target != nil {
    ...
}
```
But the condition was always getting evaluated as true. Therefore used reflection. To avoid using reflection I can think of only the following way 
```
if disk.Disk != nil {
	deviceTargetSetCount++
}
if disk.LUN != nil {
	deviceTargetSetCount++
}
if disk.CDRom != nil {
	deviceTargetSetCount++
}
```
As I understand this way is safer to use than checks using reflection, but is a bit verbose. Let me know which way to go about.

Reply:
I'd prefer the verbose option.